cmd/mrthreshold.cpp

```diff
@@ -14,222 +14,442 @@
+  + Option ("allvolumes", "compute a single threshold for all image volumes, rather than an individual threshold per volume")
 
-  + Option ("invert", "invert output binary mask.")
+  + Option ("ignorezero", "ignore zero-valued input values during threshold determination")
 
-  + Option ("toppercent", "provide a mask of the N%% top-valued voxels")
-  + Argument ("N").type_float (0.0, 100.0)
+  + Option ("mask", "compute the threshold based only on values within an input mask image")
+    + Argument ("image").type_image_in ()
 
-  + Option ("bottompercent", "provide a mask of the N%% bottom-valued voxels")
-  + Argument ("N").type_float (0.0, 100.0)
 
-  + Option ("nan", "use NaN as the output zero value.")
 
-  + Option ("ignorezero", "ignore zero-valued input voxels.")
+  + OptionGroup ("Threshold application modifiers")
+
+  + Option ("operator", "mathematical operator to use when applying the threshold; "
+                        "options are: " + join(operator_list, ",") + " (default = ge)")
+    + Argument ("choice").type_choice (operator_list)
+
+  + Option ("nan", "set voxels that fail the threshold to NaN rather than zero "
+                   "(output image will be floating-point rather than binary)");
 
-  + Option ("mask", "compute the optimal threshold based on voxels within a mask.")
-  + Argument ("image").type_image_in ();
 }
 
 
-void run ()
-{
-  default_type threshold_value (NaN), percentile (NaN), bottomNpercent (NaN), topNpercent (NaN);
-  size_t topN (0), bottomN (0), nopt (0);
+using value_type = float;
+
+
+
+bool issue_degeneracy_warning = false;
+
 
-  auto opt = get_options ("abs");
-  if (opt.size()) {
-    threshold_value = opt[0][0];
-    ++nopt;
-  }
 
-  opt = get_options ("percentile");
+Image<bool> get_mask (const Image<value_type>& in)
+{
+  Image<bool> mask;
+  auto opt = get_options ("mask");
   if (opt.size()) {
-    percentile = opt[0][0];
-    ++nopt;
+    mask = Image<bool>::open (opt[0][0]);
+    if (mask.ndim() > in.ndim())
+      throw Exception ("Cannot use mask image with more axes than input image");
+    check_dimensions (in, mask, 0, 3);
+    for (size_t axis = 3; axis != mask.ndim(); ++axis) {
+      if (mask.size (axis) > 1 && mask.size (axis) != in.size (axis))
+        throw Exception ("Dimensions of mask image do not match those of main image");
+    }
   }
+  return mask;
+}
 
-  opt = get_options ("top");
-  if (opt.size()) {
-    topN = opt[0][0];
-    ++nopt;
+
+vector<value_type> get_data (Image<value_type>& in,
+                             Image<bool>& mask,
+                             const size_t max_axis,
+                             const bool ignore_zero)
+{
+  vector<value_type> data;
+  data.reserve (voxel_count (in, 0, max_axis));
+  if (mask.valid()) {
+    Adapter::Replicate<Image<bool>> mask_replicate (mask, in);
+    if (ignore_zero) {
+      for (auto l = Loop(in, 0, max_axis) (in, mask_replicate); l; ++l) {
+        if (mask_replicate.value() && in.value() != 0.0f)
```

Comment:
Any reason we don't check if value is finite when running with `-ignore_zero`? Even if this is interpreted as using zeros to indicate NaN, if there are genuine NaNs in the image, there's no way they can be handled appropriately here, right?

Reply:
Good catch; see 05c736e.

Also revised to use `!std::isnan()` rather than `std::isfinite()`, since technically infinities can be sorted.